features/graph/api/src/main/java/org/opennms/netmgt/graph/api/generic/GenericEdge.java

```diff
@@ -125,4 +94,27 @@ public boolean equals(Object o) {
     public int hashCode() {
         return Objects.hash(super.hashCode(), source, target);
     }
+    
+    public static GenericEdgeBuilder builder() {
+        return new GenericEdgeBuilder();
+    }
+    
+    public final static class GenericEdgeBuilder extends GenericElementBuilder<GenericEdgeBuilder> {
```

Comment:
Some things I learned from the immutability work in the OIA project:
- Builders should have a private default constructor to restrict access to building them via the static builder method of the parent class.
- The build() method of the builder should pass a reference of the builder itself to the parent class constructor rather than individual arguments, this way its a bit cleaner as argument list can get long depending...
- Checks should be done on the builder's setters as well as the builders build() method to catch null values where not allowed or invalid values and throw an exception to ensure an object is never in an illegal state
- For objects that have 2 or less fields, provide a convenience method in addition to builder method to construct an immutable copy without having to type out the builder syntax
- If applicable provide a `from()` method that takes the interface type and copies it to an immutable via the builder
- Collections must be copied to an immutable copy otherwise your collections can be mutated by anyone with a reference to the original collection
- If you have collections of mutable types, you should convert them to immutable types if deep immutability is desired, this can get complicated depending on the types (see the classes here for how I handled this in OIA: https://github.com/OpenNMS/opennms-integration-api/tree/master/common/src/main/java/org/opennms/integration/api/v1/util)

Examples:
https://github.com/OpenNMS/opennms-integration-api/blob/master/common/src/main/java/org/opennms/integration/api/v1/model/immutables/ImmutableAlarm.java
https://github.com/OpenNMS/opennms-integration-api/blob/master/common/src/main/java/org/opennms/integration/api/v1/model/immutables/ImmutableEventParameter.java

Reply:
Hi @mattixtech , thanks for the feedback. These are some good points!
on 1.: makes sense, will do.
on 2.: in our case it passes a GenericXX object to the parent class constructor. I guess this has the same effect (no long list of parameters)
on 3. I am not sure I understand the benefit? Why not put the checks in the constructor of the parent class when all attributes are gathered? This gives in my opinion two benefits: no code duplication, some checks can be done in the super constructor of the parent class. A builder is anyway most of the time in an incomplete state (since we are adding until build is called).
on 4.: makes good sense to me but not sure where applicable here since we deal with a list of properties
on 5: agree
on 6: agree. This is done here https://github.com/OpenNMS/opennms/blob/d047a06016eebc677e84f53faa7bb1ac22364a68/features/graph/api/src/main/java/org/opennms/netmgt/graph/api/generic/GenericElement.java#L44
on 7. agree.

Reply:
Hey Patrick,

For #3 I just meant checks need to be done on the builders setters and also at some point when it is built. That could be done in the constructor or build method, I don't think that particularly matters. So if it is more convenient in your case to do it in the constructor that is fine.

The main thing to take away from this point is that it is useful to do the check in two places. If you have a certain field that doesn't allow nulls, the setter should check input for non-null and the build()/constructor should also check for non-null in case the setter was never called. Just doing the latter will catch illegal state but doing the former will catch illegal state sooner in a more helpful manner. Hope that makes sense.

Reply:
Also, for # 2 what I mean is that the build() method in this class for instance should look like the following:
```
public GenericEdge build() {
            return new GenericEdge(this);
        }
```

Then the constructor can be updated to take a builder as the only parameter and fields can be retrieved from the builder. I would try to do that any time you are passing more than 2 parameters to a constructor from the builder.